### PR TITLE
JSON type added for ThunderId

### PR DESCRIPTION
**Pull request by HaseenaSainul**
[DependsOn=ThunderTools:development/METROL-940]

### CODE REVIEW

File: Source/Thunder/PluginHost.cpp [hunk 1]
```diff
@@ -298,6 +298,21 @@ POP_WARNING()
...
+    PrintMetaData(const uint8_t index, const TYPE& workerId, const uint32_t runs)
+    {
+        printf("  Thread%02d|0x%16" PRIxPTR ": %10d", (index), reinterpret_cast<uintptr_t>(workerId), runs);
+    }
+
+    template <typename TYPE>
+    inline typename std::enable_if<(!std::is_same<TYPE, void*>::value), void>::type
+    PrintMetaData(const uint8_t index, const TYPE workerId, const uint32_t runs)
+    {
+        printf("  Thread%02d|0x%16" PRIu64 ": %10d", (index), static_cast<uint64_t>(workerId), runs);
```
Comment (sebaszm) [inline Source/Thunder/PluginHost.cpp:312]: This is DWORD on Windows, so no need to go through uint64_t.

File: Source/core/JSON.h [hunk 2]
```diff
@@ -4980,6 +4979,114 @@ namespace Core {
...
+            }
+
+            template <typename TYPE, typename std::enable_if<!(std::is_same<TYPE, void*>::value)>::type>
+            ThreadId(const TYPE id)
+                : JSON::String(false)
+            {
+                String::operator=(Core::NumberType<TYPE, false, NumberBase::BASE_HEXADECIMAL>(id).Text());
+            }
+
+            ThreadId(const ThreadId&& move)
```
Comment (sebaszm) [inline Source/core/JSON.h:5052]: moved object must not be const to be moved

File: Source/core/JSON.h [hunk 3]
```diff
@@ -4980,6 +4979,114 @@ namespace Core {
...
+            Default() const
+            {
+                return reinterpret_cast<thread_id>(Core::NumberType<uintptr_t, false, NumberBase::BASE_HEXADECIMAL>(String::Default().c_str(), static_cast<uint32_t>(String::Default().length())).Value());
+            }
+
+            template <typename TYPE>
+            inline typename std::enable_if<!(std::is_same<TYPE, void*>::value), thread_id>::type
+            Default() const
+            {
+                return static_cast<thread_id>(Core::NumberType<uint64_t, false, NumberBase::BASE_HEXADECIMAL>(String::Default().c_str(), static_cast<uint32_t>(String::Default().length())).Value());
```
Comment (sebaszm) [inline Source/core/JSON.h:5103]: Again it doesnt need to be as big as uint64_t

File: Source/core/JSON.h [hunk 4]
```diff
@@ -4980,6 +4979,114 @@ namespace Core {
...
+            }
+
+            template <typename TYPE>
+            inline typename std::enable_if<!(std::is_same<TYPE, void*>::value), thread_id>::type
+            Value() const
+            {
+                return static_cast<thread_id>(Core::NumberType<uint64_t, false, NumberBase::BASE_HEXADECIMAL>(String::Value().c_str(), static_cast<uint32_t>(String::Value().length())).Value());
+            }
+
+            inline thread_id Value() const
```
Comment (sebaszm) [inline Source/core/JSON.h:5125]: ::thread_id

File: Source/core/JSON.h [hunk 5]
```diff
@@ -4980,6 +4979,114 @@ namespace Core {
...
+                return Value<thread_id>();
+            }
+
+            inline operator thread_id() const
+            {
+                return Value();
+            }
+        };
+
+        typedef ThreadId Pointer;
```
Comment (sebaszm) [inline Source/core/JSON.h:5136]: Perhaps it's better to remove the Pointer typedef all along? As it's either gonna be InstanceId or ThreadId and update the code that's using it accordingly.

File: Source/core/Portability.h [hunk 6]
```diff
@@ -680,9 +680,9 @@ typedef std::string string;
 #define EMPTY_STRING _T("")
 
 #ifdef __LINUX__
-typedef pthread_t ThreadId;
+typedef pthread_t thread_id;
```
Comment (sebaszm) [inline Source/core/Portability.h:683]: Why not Core::thread_id? Consistenly with Core::instance_id?

File: Source/plugins/IController.h [hunk 7]
```diff
@@ -270,7 +270,7 @@ namespace Controller {
             };
 
             struct Thread {
-                Core::instance_id Id /* @brief Thread ID */;
+                ::thread_id Id /* @brief Thread ID */;
```
Comment (sebaszm) [inline Source/plugins/IController.h:273]: This means that IMetadata can't go over COM-RPC.

File: Source/plugins/Metadata.cpp [hunk 8]
```diff
@@ -241,7 +241,7 @@ namespace PluginHost
         Add(_T("runs"), &Runs);
     }
     Metadata::Server::Minion& Metadata::Server::Minion::operator=(const Core::ThreadPool::Metadata& info) {
-        Id = (Core::instance_id)info.WorkerId;
+        Id = (::thread_id)info.WorkerId;
```
Comment (sebaszm) [inline Source/plugins/Metadata.cpp:244]: There seems to be a lot of C-style casting to ::thread_id throughout code...

File: Source/core/Thread.cpp [hunk 9]
```diff
@@ -132,14 +132,14 @@ namespace Core {
 #endif
     }
 
-    ::ThreadId Thread::ThreadId()
+    thread_id Thread::ThreadId()
     {
 #ifdef __WINDOWS__
 PUSH_WARNING(DISABLE_WARNING_CONVERSION_TO_GREATERSIZE)
```
Comment (sebaszm) [inline Source/core/Thread.cpp:138]: thread_id is DWORD and GetCurrentThreadId() return DWORD, why would there be a warning?

**Review by sebaszm — CHANGES_REQUESTED** [review]

**Comment by HaseenaSainul** [comment]
There is no permission to update the branch in rdkcentral, so new PR created from fork: https://github.com/rdkcentral/Thunder/pull/1766.
Hence closing this PR.

**Pull request closed without merging**